docker/pandoc-lambda/function/app.py

```diff
@@ -1,14 +1,133 @@
 import boto3
+import io
 import os
 import signal
 import subprocess
 import tempfile
+from docx import Document
+from docx.oxml import OxmlElement, parse_xml
+from lxml import etree
+
+def lift_footnote(doc, footnotes_part, ref, texts, id, author=False):
+    id_att = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id'
+    custom_mark_att = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}customMarkFollows'
+    footnote = OxmlElement("w:footnote")
+    footnote.attrib[id_att] = id
+    doc_insert = OxmlElement("w:footnoteReference")
+    doc_insert.attrib[custom_mark_att] = "1"
+    doc_insert.attrib[id_att] = id
+
+    # Content
+    for t in texts:
+        footnote.append(t)
+
+    # Insert into the footnotes file
+    footnotes_part.element.append(footnote)
+
+    # Insert the reference into the doc
+    ref.insert(1,doc_insert)
+
+def promote_case_footnotes(doc):
+    val_att = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val'
+    author_footnotes = {}
+
+    for ref in doc.element.xpath("//*[*/w:rStyle[starts-with(@w:val,'FootnoteReference')]]"):
+        mark_text = ref.text
+        style_node = ref.xpath(".//w:rStyle[starts-with(@w:val,'FootnoteReference')]")[0]
+        node_id = style_node.attrib[val_att][18:]
+        mark_id = f"{node_id}-{mark_text}"
+        if mark_id not in author_footnotes:
```

Comment:
Q: if we encounter `mark_id in author_footnotes`, does that mean there's an infelicity that should be addressed by the author (e.g. choosing a non-duplicative footnote mark), by us (adapting the html code so it doesn't produce duplicative footnote marks), etc. or is this a benign situation? Should we figure out a way to log duplicative notes that are skipped via this branch?

Reply:
This could happen in the case that an author has annotated their own text (possibly cloned), and be benign. We still capture+output the note if it's a duplicate. Logging might be a good idea, or I've been thinking about adding in a fab task for bulk analysis when caching is implemented.

Reply:
> We still capture+output the note if it's a duplicate.
Ah! I was misreading, yes, now I see. Thanks!

Love the idea of using the cache, once implemented, for bulk analysis of varying kinds!